services/api/service.go

```diff
@@ -1748,17 +1754,21 @@ func (api *RelayAPI) checkSubmissionSlotDetails(w http.ResponseWriter, log *logr
 			api.RespondError(w, http.StatusBadRequest, "no duty found for submission slot")
 			return false
 		}
-
 		if duty.IsMevCommitValidator {
-			isBuilderRegistered, err := api.datastore.IsMevCommitBlockBuilder(common.NewPubkeyHex(submission.BidTrace.BuilderPubkey.String()))
-			if err != nil {
-				log.WithError(err).Error("Failed to check builder registration")
-				api.RespondError(w, http.StatusInternalServerError, "Internal server error")
-				return false
-			}
-			if !isBuilderRegistered {
-				// TODO: Implement caching strategy for builder registration status
-				api.RespondError(w, http.StatusBadRequest, "Builder pubkey is not registered under mev-commit")
+			builderCacheEntry, ok := api.blockBuildersCache[submission.BidTrace.BuilderPubkey.String()]
+			if !ok {
+				isBuilderRegistered, err := api.datastore.IsMevCommitBlockBuilder(common.NewPubkeyHex(submission.BidTrace.BuilderPubkey.String()))
+				if err != nil {
+					log.WithError(err).Error("Failed to check builder registration")
+					api.RespondError(w, http.StatusInternalServerError, "Internal server error")
+					return false
+				}
+				if !isBuilderRegistered {
+					api.RespondError(w, http.StatusBadRequest, "Builder pubkey is not registered under mev-commit")
+					return false
+				}
+			} else if !builderCacheEntry.mevCommitOptInStatus {
+				api.RespondError(w, http.StatusBadRequest, "Builder pubkey is not opted into mev-commit")
```

Comment:
I would suggest to move this check into separated function to decrease amount of nested if's and improve readability:

```go
func (api *API) validateDuty(w http.ResponseWriter, duty Duty, submission Submission) bool {
    if !duty.IsMevCommitValidator {
        return true
    }

    builderPubKeyStr := submission.BidTrace.BuilderPubkey.String()

    if builderCacheEntry, ok := api.blockBuildersCache[builderPubKeyStr]; ok {
        if !builderCacheEntry.mevCommitOptInStatus {
            api.RespondError(w, http.StatusBadRequest, "Builder pubkey is not opted into mev-commit")
            return false
        }
        return true
    }

    isBuilderRegistered, err := api.datastore.IsMevCommitBlockBuilder(common.NewPubkeyHex(builderPubKeyStr))
    if err != nil {
        log.WithError(err).Error("Failed to check builder registration")
        api.RespondError(w, http.StatusInternalServerError, "Internal server error")
        return false
    }
    if !isBuilderRegistered {
        api.RespondError(w, http.StatusBadRequest, "Builder pubkey is not registered under mev-commit")
        return false
    }

    return true
}

// In your checkSubmissionSlotDetails function
if !api.validateDuty(w, duty, submission) {
    return false
}
duration := time.Since(start)
log.WithField("duration", duration).Trace("mev-commit check completed")
```